File: face/geomajas-face-dojo/client/src/main/resources/geomajas/map/print/template/DefaultTemplateBuilder.js
dojo.provide("geomajas.map.print.template.DefaultTemplateBuilder");
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
dojo.declare("DefaultTemplateBuilder", TemplateBuilder, {

	/**
	 * @class 
	 * Default print template builder, parameters include title, size, raster DPI, orientation, etc...
	 * 
	 * @author Jan De Moerloose
	 *
	 * @constructor
	 */
	constructor : function () {
//		private double pageWidth;
//		private double pageHeight;
//		private int marginX;
//		private int marginY;
//		private String titleText;
//		private int rasterDpi;
//		private boolean withScaleBar;
//		private boolean withArrow;
//		private MapModel mapModel;
//		private String applicationId;
	},
	
	buildTemplate : function () {
		var template = this.inherited(arguments);
		template.setTemplate(false);
		template.setId(1);
		template.setName("default");
		return template;
	},	
	
	buildPage : function () {
		var page = this.inherited(arguments);
		page.getLayoutConstraint().setWidth(this.pageWidth);
		page.getLayoutConstraint().setHeight(this.pageHeight);
		return page;
	},
	
	buildMap : function () {
		var map = this.inherited(arguments);
		map.getLayoutConstraint().setMarginX(this.marginX);
		map.getLayoutConstraint().setMarginY(this.marginY);
		var view = this.mapModel.getMapView();
		var mapWidth = this.pageWidth - 2 * this.marginX;
		var mapHeight = this.pageHeight - 2 * this.marginY;
		var bounds = view.getCurrentBbox().createFittingBox(mapWidth, mapHeight);
		var origin = bounds.getOrigin();
		map.setLocation(new Coordinate(origin.getX(), origin.getY()));
		map.setPpUnit((mapWidth / bounds.getWidth()));
		map.setTag("map");
		map.setMapId(this.mapModel.getId());
		map.setApplicationId(this.applicationId);
		map.setRasterResolution(this.rasterDpi);
		var layerChildren = map.getChildren().list; /*PrintComponentInfo*/
		
		var layers = this.mapModel.getVisibleLayerList();
		for(var i = 0; i < layers.count; i++){
			var layer = layers.item(i);
			if (layer instanceof VectorLayer && layer.checkVisibility()) {
				var info = new VectorLayerComponentInfo();
				var vectorLayer = layer;
				info.setLayerId(vectorLayer.getLayerId());
				info.setStyleInfo(vectorLayer.getNamedStyle().toJSON());
				info.setFilter(vectorLayer.getFilterString());
				info.setLabelsVisible(vectorLayer.isLabeled());
				info.setSelected(vectorLayer.isSelected());
				info.setSelectedFeatureIds(vectorLayer.getSelectionStore().getKeyList());
				layerChildren.splice(0,0,info);
			} else if (layer instanceof RasterLayer && layer.checkVisibility()) {
				var info = new RasterLayerComponentInfo();
				var rasterLayer = layer;
				info.setLayerId(rasterLayer.getLayerId());
				info.setStyle(rasterLayer.getStyle().opacity);
				layerChildren.splice(0,0,info);
			}
		}
		return map;
	},
	
	buildArrow : function () {
		if (this.withArrow == true) {
			var northarrow = this.inherited(arguments);
			northarrow.setImagePath("/images/northarrow.gif");
			northarrow.getLayoutConstraint().setAlignmentX(geomajas.LayoutConstraints.RIGHT);
			northarrow.getLayoutConstraint().setAlignmentY(geomajas.LayoutConstraints.TOP);
			northarrow.getLayoutConstraint().setMarginX(20);
			northarrow.getLayoutConstraint().setMarginY(20);
			northarrow.getLayoutConstraint().setWidth(10);
			northarrow.setTag("arrow");
			return northarrow;
		} else {
			return null;
		}
	},
	
	buildLegend : function () {
		var legend = this.inherited(arguments);
		var style = new FontStyleInfo();
		style.setFamily("Dialog");
		style.setStyle("Italic");
		style.setSize(14);
		legend.setFont(style);
		legend.setMapId(this.mapModel.getId());
		legend.setTag("legend");
		
		var layers = this.mapModel.getVisibleLayerList();
		for(var i = 0; i < layers.count; i++){
			var layer = layers.item(i);
			if (layer instanceof VectorLayer && layer.checkVisibility()) {
				var vectorLayer = layer;
				var defs = vectorLayer.getStyles();
				for(var j = 0; j < defs.count; j++){
					var styleDefinition = defs.item(j);
					var text = "";
					if (defs.count > 1) {
						text = vectorLayer.getLabel() + " (" + styleDefinition.getName() + ")";
					} else {
						text = vectorLayer.getLabel();
					}
					var item = new LegendItemComponentInfo();
					var icon = new LegendIconComponentInfo();
					icon.setLabel(text);
					icon.setStyleInfo(styleDefinition.toJSON());
					icon.setLayerType(vectorLayer.getLayerType());
					var legendLabel = new LabelComponentInfo();
					legendLabel.setBackgroundColor("0xFFFFFF");
					legendLabel.setBorderColor("0x000000");
					legendLabel.setFontColor("0x000000");
					legendLabel.setFont(legend.getFont());
					legendLabel.setText(text);
					legendLabel.setTextOnly(true);
					item.addChild(icon);
					item.addChild(legendLabel);
					legend.addChild(item);
				}
			} else if (layer instanceof RasterLayer && layer.checkVisibility()) {
				var rasterLayer = layer;
				var item = new LegendItemComponentInfo();
				var icon = new LegendIconComponentInfo();
				icon.setLabel(rasterLayer.getLabel());
				icon.setLayerType(rasterLayer.getLayerType());
				var legendLabel = new LabelComponentInfo();
				legendLabel.setFont(legend.getFont());
				legendLabel.setBackgroundColor("0xFFFFFF");
				legendLabel.setBorderColor("0x000000");
				legendLabel.setFontColor("0x000000");
				legendLabel.setText(rasterLayer.getLabel());
				legendLabel.setTextOnly(true);
				item.addChild(icon);
				item.addChild(legendLabel);
				legend.addChild(item);
			}
		}
		return legend;
	},

	buildScaleBar : function () {
		if (this.withScaleBar == true) {
			var bar = this.inherited(arguments);
			bar.setTicNumber(3);
			bar.setTag("scalebar");
			return bar;
		} else {
			return null;
		}
	},

	buildTitle : function () {
		if (this.titleText != null) {
			var title = this.inherited(arguments);
			title.setText(this.titleText);
			title.getLayoutConstraint().setMarginY(2 * this.marginY);
			return title;
		} else {
			return null;
		}
	},
	
	// -----------------------------------------------------
	
	setPageWidth : function (pageWidth) {
		this.pageWidth = pageWidth;
	},

	setPageHeight : function (pageHeight) {
		this.pageHeight = pageHeight;
	},

	setTitleText : function (titleText) {
		this.titleText = titleText;
	},

	setRasterDpi : function (rasterDpi) {
		this.rasterDpi = rasterDpi;
	},

	setWithScaleBar : function (withScaleBar) {
		this.withScaleBar = withScaleBar;
	},

	setWithArrow : function (withArrow) {
		this.withArrow = withArrow;
	},

	setMarginX : function (marginX) {
		this.marginX = marginX;
	},

	setMarginY : function (marginY) {
		this.marginY = marginY;
	},

	setMapModel : function (mapModel) {
		this.mapModel = mapModel;
	},

	setApplicationId : function (applicationId) {
		this.applicationId = applicationId;
	}

});

File: plugin/geomajas-plugin-printing/printing-gwt/src/main/java/org/geomajas/plugin/printing/client/template/DefaultTemplateBuilder.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2012 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
package org.geomajas.plugin.printing.client.template;

import java.util.ArrayList;
import java.util.List;

import org.geomajas.configuration.FontStyleInfo;
import org.geomajas.configuration.client.ClientLayerInfo;
import org.geomajas.configuration.client.ClientRasterLayerInfo;
import org.geomajas.configuration.client.ClientVectorLayerInfo;
import org.geomajas.geometry.Coordinate;
import org.geomajas.gwt.client.map.MapModel;
import org.geomajas.gwt.client.map.MapView;
import org.geomajas.gwt.client.map.layer.Layer;
import org.geomajas.gwt.client.map.layer.RasterLayer;
import org.geomajas.gwt.client.map.layer.VectorLayer;
import org.geomajas.plugin.printing.client.util.PrintingLayout;
import org.geomajas.plugin.printing.command.dto.PrintTemplateInfo;
import org.geomajas.plugin.printing.component.dto.ImageComponentInfo;
import org.geomajas.plugin.printing.component.dto.LabelComponentInfo;
import org.geomajas.plugin.printing.component.dto.LayoutConstraintInfo;
import org.geomajas.plugin.printing.component.dto.LegendComponentInfo;
import org.geomajas.plugin.printing.component.dto.LegendGraphicComponentInfo;
import org.geomajas.plugin.printing.component.dto.LegendIconComponentInfo;
import org.geomajas.plugin.printing.component.dto.LegendItemComponentInfo;
import org.geomajas.plugin.printing.component.dto.MapComponentInfo;
import org.geomajas.plugin.printing.component.dto.PageComponentInfo;
import org.geomajas.plugin.printing.component.dto.PrintComponentInfo;
import org.geomajas.plugin.printing.component.dto.RasterLayerComponentInfo;
import org.geomajas.plugin.printing.component.dto.RasterizedLayersComponentInfo;
import org.geomajas.plugin.printing.component.dto.ScaleBarComponentInfo;
import org.geomajas.plugin.rasterizing.command.dto.RasterLayerRasterizingInfo;
import org.geomajas.sld.FeatureTypeStyleInfo;
import org.geomajas.sld.RuleInfo;

/**
 * Default print template builder, parameters include title, size, raster DPI, orientation, etc...
 * 
 * @author Jan De Moerloose
 */
public class DefaultTemplateBuilder extends AbstractTemplateBuilder {

	protected double pageWidth;

	protected double pageHeight;

	protected int marginX;

	protected int marginY;

	protected String titleText;

	protected int rasterDpi;

	protected boolean withScaleBar;

	protected boolean withArrow;

	protected MapModel mapModel;

	protected String applicationId;

	@Override
	public PrintTemplateInfo buildTemplate() {
		PrintTemplateInfo template = super.buildTemplate();
		template.setId(1L);
		template.setName("default");
		return template;
	}

	@Override
	protected PageComponentInfo buildPage() {
		PageComponentInfo page = super.buildPage();
		page.getLayoutConstraint().setWidth((float) pageWidth);
		page.getLayoutConstraint().setHeight((float) pageHeight);
		return page;
	}

	@Override
	protected MapComponentInfo buildMap() {
		MapComponentInfo map = super.buildMap();
		map.getLayoutConstraint().setMarginX(marginX);
		map.getLayoutConstraint().setMarginY(marginY);
		MapView view = mapModel.getMapView();
		double mapWidth = getPageWidth() - 2 * marginX;
		double mapHeight = getPageHeight() - 2 * marginY;
		Coordinate origin = view.getBounds().createFittingBox(mapWidth, mapHeight).getOrigin();
		map.setLocation(new org.geomajas.geometry.Coordinate(origin.getX(), origin.getY()));
		map.setPpUnit((float) (mapWidth / view.getBounds().createFittingBox(mapWidth, mapHeight).getWidth()));
		map.setTag("map");
		map.setMapId(mapModel.getMapInfo().getId());
		map.setApplicationId(applicationId);
		map.setRasterResolution(rasterDpi);
		List<PrintComponentInfo> layers = new ArrayList<PrintComponentInfo>();
		// use the normal way for raster layers (TODO: add support for dpi to rasterized part)
		for (Layer layer : mapModel.getLayers()) {
			if (layer instanceof RasterLayer && layer.isShowing()) {
				RasterLayerComponentInfo info = new RasterLayerComponentInfo();
				RasterLayer rasterLayer = (RasterLayer) layer;
				info.setLayerId(rasterLayer.getServerLayerId());
				info.setStyle(rasterLayer.getLayerInfo().getStyle());
				layers.add(info);
			}
		}
		// use the rasterized layers way for vector layers
		for (ClientLayerInfo layerInfo : mapModel.getMapInfo().getLayers()) {
			// we must skip the raster layers or we have them twice !
			if (layerInfo instanceof ClientRasterLayerInfo) {
				RasterLayerRasterizingInfo rInfo = (RasterLayerRasterizingInfo) layerInfo
						.getWidgetInfo(RasterLayerRasterizingInfo.WIDGET_KEY);
				rInfo.setShowing(false);
			}
		}
		RasterizedLayersComponentInfo rasterizedLayersComponentInfo = new RasterizedLayersComponentInfo();
		rasterizedLayersComponentInfo.setMapInfo(mapModel.getMapInfo());
		layers.add(rasterizedLayersComponentInfo);
		map.getChildren().addAll(0, layers);
		return map;
	}

	@Override
	protected ImageComponentInfo buildArrow() {
		if (isWithArrow()) {
			ImageComponentInfo northarrow = super.buildArrow();
			northarrow.setImagePath("/images/northarrow.gif");
			northarrow.getLayoutConstraint().setAlignmentX(LayoutConstraintInfo.RIGHT);
			northarrow.getLayoutConstraint().setAlignmentY(LayoutConstraintInfo.TOP);
			northarrow.getLayoutConstraint().setMarginX((float) PrintingLayout.templateMarginX);
			northarrow.getLayoutConstraint().setMarginY((float) PrintingLayout.templateMarginY);
			northarrow.getLayoutConstraint().setWidth((float) PrintingLayout.templateNorthArrowWidth);
			northarrow.setTag("arrow");
			return northarrow;
		} else {
			return null;
		}
	}

	@Override
	protected LegendComponentInfo buildLegend() {
		LegendComponentInfo legend = super.buildLegend();
		FontStyleInfo style = new FontStyleInfo();
		style.setFamily(PrintingLayout.templateDefaultFontFamily);
		style.setStyle(PrintingLayout.templateDefaultFontStyle);
		style.setSize((int) PrintingLayout.templateDefaultFontSize);
		legend.setFont(style);
		legend.setMapId(mapModel.getMapInfo().getId());
		legend.setTag("legend");
		for (Layer layer : mapModel.getLayers()) {
			if (layer instanceof VectorLayer && layer.isShowing()) {
				VectorLayer vectorLayer = (VectorLayer) layer;
				ClientVectorLayerInfo layerInfo = vectorLayer.getLayerInfo();
				String label = layerInfo.getLabel();
				FeatureTypeStyleInfo fts = layerInfo.getNamedStyleInfo().getUserStyle().getFeatureTypeStyleList().get(0);
				for (RuleInfo rule : fts.getRuleList()) {
					// use title if present, name if not
					String title = (rule.getTitle() != null ? rule.getTitle() : rule.getName());
					// fall back to style name
					if (title == null) {
						title = layerInfo.getNamedStyleInfo().getName();
					}
					LegendItemComponentInfo item = new LegendItemComponentInfo();
					LegendGraphicComponentInfo graphic = new LegendGraphicComponentInfo();
					graphic.setLabel(title);
					graphic.setRuleInfo(rule);
					graphic.setLayerId(layerInfo.getServerLayerId());
					item.addChild(graphic);
					item.addChild(getLegendLabel(legend, title));
					legend.addChild(item);
				}
			} else if (layer instanceof RasterLayer && layer.isShowing()) {
				RasterLayer rasterLayer = (RasterLayer) layer;
				ClientRasterLayerInfo layerInfo = rasterLayer.getLayerInfo();
				LegendItemComponentInfo item = new LegendItemComponentInfo();
				LegendIconComponentInfo icon = new LegendIconComponentInfo();
				icon.setLabel(layerInfo.getLabel());
				icon.setLayerType(layerInfo.getLayerType());
				item.addChild(icon);
				item.addChild(getLegendLabel(legend, layerInfo.getLabel()));
				legend.addChild(item);
			}
		}
		return legend;
	}

	private LabelComponentInfo getLegendLabel(LegendComponentInfo legend, String text) {
		LabelComponentInfo legendLabel = new LabelComponentInfo();
		legendLabel.setBackgroundColor(PrintingLayout.templateDefaultBackgroundColor);
		legendLabel.setBorderColor(PrintingLayout.templateDefaultBorderColor);
		legendLabel.setFontColor(PrintingLayout.templateDefaultColor);
		legendLabel.setFont(legend.getFont());
		legendLabel.setText(text);
		legendLabel.setTextOnly(true);
		return legendLabel;
	}

	@Override
	protected ScaleBarComponentInfo buildScaleBar() {
		if (isWithScaleBar()) {
			ScaleBarComponentInfo bar = super.buildScaleBar();
			bar.setTicNumber(3);
			bar.setTag("scalebar");
			return bar;
		} else {
			return null;
		}
	}

	@Override
	protected LabelComponentInfo buildTitle() {
		if (titleText != null) {
			LabelComponentInfo title = super.buildTitle();
			title.setText(titleText);
			title.getLayoutConstraint().setMarginY(2 * marginY);
			return title;
		} else {
			return null;
		}
	}

	public double getPageWidth() {
		return pageWidth;
	}

	public void setPageWidth(double pageWidth) {
		this.pageWidth = pageWidth;
	}

	public double getPageHeight() {
		return pageHeight;
	}

	public void setPageHeight(double pageHeight) {
		this.pageHeight = pageHeight;
	}

	public String getTitleText() {
		return titleText;
	}

	public void setTitleText(String titleText) {
		this.titleText = titleText;
	}

	public int getRasterDpi() {
		return rasterDpi;
	}

	public void setRasterDpi(int rasterDpi) {
		this.rasterDpi = rasterDpi;
	}

	public boolean isWithScaleBar() {
		return withScaleBar;
	}

	public void setWithScaleBar(boolean withScaleBar) {
		this.withScaleBar = withScaleBar;
	}

	public boolean isWithArrow() {
		return withArrow;
	}

	public void setWithArrow(boolean withArrow) {
		this.withArrow = withArrow;
	}

	public int getMarginX() {
		return marginX;
	}

	public void setMarginX(int marginX) {
		this.marginX = marginX;
	}

	public int getMarginY() {
		return marginY;
	}

	public void setMarginY(int marginY) {
		this.marginY = marginY;
	}

	public MapModel getMapModel() {
		return mapModel;
	}

	public void setMapModel(MapModel mapModel) {
		this.mapModel = mapModel;
	}

	public String getApplicationId() {
		return applicationId;
	}

	public void setApplicationId(String applicationId) {
		this.applicationId = applicationId;
	}

}
